### PR TITLE
Add Feature Flag to Disable Array Preallocation for Log Record Attributes in Limited Stack Environments

### LINKED ISSUE #1920: LogRecord Attributes - avoid heap allocations
We recently decided to [pre-allocate up to 8](https://github.com/open-telemetry/opentelemetry-rust/pull/1908) attributes, to avoid reallocation/copy costs, but attributes are still always on the heap. 
Opening an issue to track improving this, so that, up to 8 attributes can be stack allocated, and then fallback to heap for more attributes. This, along with https://github.com/open-telemetry/opentelemetry-rust/issues/1919, allow one to write logs without any allocation when using synchronous exporters like etw/user_events that require no batching.

We could add a dependency on [SmallVec](https://docs.rs/smallvec/) or make a home-grown solution. For an example implementation, [.NET's `TagList`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/TagList.netcore.cs) can be referred, which optimizes for up to 8 attributes.

We may need a feature flag to control this, as some scenarios (eg: Kernel mode) has less stack available.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1920 

## Changes
This PR introduces a feature flag, `no-preallocate-attributes-array`, that disables the default behavior of preallocating a fixed-size array for storing log record attributes. Instead of using a preallocated array, this feature flag enables the use of the vector part of `GrowableArray`, which allocates memory dynamically when the first entry is added.

Rationale: Kernel mode and other environments with limited stack space require careful management of stack allocations. By using a dynamically growing vector instead of preallocating a fixed-size array, we can reduce the stack footprint.

Note - We could have directly used the `Vector` instead of `GrowableArray` for this scenario, however didn't find any visible difference in stress tests with either approach, so kept `GrowableArray` to be consistent.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-sdk/Cargo.toml [hunk 1]
```diff
@@ -51,6 +51,7 @@ testing = ["opentelemetry/testing", "trace", "metrics", "logs", "rt-async-std",
 rt-tokio = ["tokio", "tokio-stream"]
 rt-tokio-current-thread = ["tokio", "tokio-stream"]
 rt-async-std = ["async-std"]
+no-preallocate-attributes-array = []
```
Comment (cijothomas) [inline opentelemetry-sdk/Cargo.toml:54]: Alternate suggestion:
`disable_stack_log_attributes`

"Preallocate" is potentially misleading, we can pre-allocate in heap as well.

Comment (cijothomas) [reply opentelemetry-sdk/Cargo.toml:54]: please also add a changelog entry for this. There is a general lacking of documentaion for each of the feature flags, but that is something we need to work on independent of this PR.

**Comment by utpilla** [comment]
> Rationale: Kernel mode and other environments with limited stack space require careful management of stack allocations. By using a dynamically growing vector instead of preallocating a fixed-size array, we can reduce the stack footprint.

Do we have users who have tried to run OpenTelemetry Rust in Kernel mode? I'm asking mainly to understand if the lack of this feature is an actual blocker for anyone.

**Comment by lalitb** [comment]
> Do we have users who have tried to run OpenTelemetry Rust in Kernel mode? 

We don't yet support kernel mode, because of usage of `std` constructs. That needs to be addressed separately. Irrespective of that, kernel stack size is limited (8kb in x32 and 16 kb in x64) for Linux kernel, so this feature-flag would be relevant once the kernel support is there.

**Comment by cijothomas** [comment]
> > Do we have users who have tried to run OpenTelemetry Rust in Kernel mode?
> 
> We don't yet support kernel mode, because of usage of `std` constructs. That needs to be addressed separately. Irrespective of that, kernel stack size is limited (8kb in x32 and 16 kb in x64) for Linux kernel, so this feature-flag would become relevant once the kernel support is there.

Is it better to delay adding this feature flag until we reach that state? Just to keep overall feature-flags under control.
We can have a tracking issue to gauge the overall interest in the feature (there could be other scenarios beyond kernel mode code)

**Comment by utpilla** [comment]
> Is it better to delay adding this feature flag until we reach that state?

I would be in favor of that.

> Irrespective of that, kernel stack size is limited (8kb in x32 and 16 kb in x64) for Linux kernel, so this feature-flag would become relevant once the kernel support is there.

It'd be good to try it out first and check if we run into issues. (once we start supporting this scenario).

We allocate 10 instances of `Option<(Key, AnyValue)>` on stack. The size of `Option<(Key, AnyValue)>` should be around 48 bytes. 480 bytes of stack allocation for a stack size of 8kb/16kb is probably okay?

**Comment by lalitb** [comment]
> Is it better to delay adding this feature flag until we reach that state?

sure, I actually prefer that. The feature flag requirement was added in #1920, so create the PR :)

> We allocate 10 instances of Option<(Key, AnyValue)> on stack. The size of Option<(Key, AnyValue)> should be around 48 bytes. 480 bytes of stack allocation for a stack size of 8kb/16kb is probably okay?

I think tracing needs to be conservative for any stack allocation in kernel mode, specifically these pre-allocations in stack. Thinking about the deep call chains, and recursion in actual kernel mode application, the stack size can grow fast. But, there would be other places we need to look too, and good to revisit once we think about kernel-mode :)

**Comment by lalitb** [comment]
closing as discussed above.

**Pull request closed without merging**